File: claude-code-agents-python/src/orchestrator/context_injector.py
"""Context injection for HeadlessRunner prompts."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from ..models.crawl_target import CrawlTarget
    from .improvement_chain import ImprovementChain


def inject_context(
    iteration: int,
    chain: Optional[ImprovementChain] = None,
    target: Optional[CrawlTarget] = None,
) -> str:
    """Generate a context fragment to inject into HeadlessRunner prompts.

    Aggregates information from the improvement chain history and current
    target to provide steering guidance for the next crawl iteration.

    Args:
        iteration: Current iteration number (0-based).
        chain: The improvement chain with historical deltas.
        target: The current crawl target.

    Returns:
        A prompt fragment string with accumulated context.
    """
    sections: list[str] = []

    sections.append(f"## Iteration Context (iteration {iteration})")

    if chain and chain.iteration_count > 0:
        cumulative = chain.get_cumulative_delta()
        if cumulative:
            sections.append(
                f"### Quality Trajectory\n"
                f"- Starting quality: {cumulative.quality_before:.3f}\n"
                f"- Current quality: {cumulative.quality_after:.3f}\n"
                f"- Total improvement: {chain.total_improvement:+.3f}\n"
                f"- Iterations completed: {chain.iteration_count}"
            )

            if cumulative.new_patterns:
                patterns_str = "\n".join(f"  - `{p}`" for p in cumulative.new_patterns[:10])
                sections.append(f"### Discovered Patterns\n{patterns_str}")

            if cumulative.failing_selectors:
                failing_str = "\n".join(
                    f"  - `{s}`" for s in cumulative.failing_selectors[:10]
                )
                sections.append(
                    f"### Failing Selectors (avoid these)\n{failing_str}"
                )

            if cumulative.discovered_page_types:
                types_str = ", ".join(cumulative.discovered_page_types)
                sections.append(f"### Discovered Page Types\n{types_str}")

            if cumulative.steer_direction:
                sections.append(
                    f"### Steering Direction\n{cumulative.steer_direction}"
                )

        history = chain.get_history()
        if len(history) >= 2:
            recent = history[-1]
            sections.append(
                f"### Last Iteration Summary\n"
                f"- Quality change: {recent.quality_improvement:+.3f}\n"
                f"- New patterns found: {len(recent.new_patterns)}\n"
                f"- Selectors failing: {len(recent.failing_selectors)}"
            )
    else:
        sections.append(
            "This is the initial crawl iteration. "
            "Focus on broad extraction and discovery."
        )

    if target:
        sections.append(
            f"### Current Target\n"
            f"- URL: {target.url}\n"
            f"- Spider: {target.spider_name}\n"
            f"- Max pages: {target.max_pages}\n"
            f"- Priority: {target.priority}"
        )
        if target.page_type_hint:
            sections.append(f"- Expected page type: {target.page_type_hint.value}")

    return "\n\n".join(sections)

File: claude-code-agents-typescript/src/orchestrator/context-injector.ts
// src/orchestrator/context-injector.ts — Context injection for HeadlessRunner prompts
import type { CrawlTarget } from '../models/crawl-target.js';
import type { ImprovementChain } from './improvement-chain.js';

export function injectContext(
  iteration: number,
  chain?: ImprovementChain,
  target?: CrawlTarget,
): string {
  const sections: string[] = [];

  sections.push(`## Iteration Context (iteration ${iteration})`);

  if (chain && chain.iterationCount > 0) {
    const cumulative = chain.getCumulativeDelta();
    if (cumulative) {
      sections.push(
        [
          '### Quality Trajectory',
          `- Starting quality: ${(cumulative.qualityBefore as number).toFixed(3)}`,
          `- Current quality: ${(cumulative.qualityAfter as number).toFixed(3)}`,
          `- Total improvement: ${chain.totalImprovement >= 0 ? '+' : ''}${chain.totalImprovement.toFixed(3)}`,
          `- Iterations completed: ${chain.iterationCount}`,
        ].join('\n'),
      );

      if (cumulative.newPatterns.length > 0) {
        const patternsStr = cumulative.newPatterns
          .slice(0, 10)
          .map((p) => `  - \`${p}\``)
          .join('\n');
        sections.push(`### Discovered Patterns\n${patternsStr}`);
      }

      if (cumulative.failingSelectors.length > 0) {
        const failingStr = cumulative.failingSelectors
          .slice(0, 10)
          .map((s) => `  - \`${s}\``)
          .join('\n');
        sections.push(
          `### Failing Selectors (avoid these)\n${failingStr}`,
        );
      }

      if (cumulative.discoveredPageTypes.length > 0) {
        sections.push(
          `### Discovered Page Types\n${cumulative.discoveredPageTypes.join(', ')}`,
        );
      }

      if (cumulative.steerDirection) {
        sections.push(
          `### Steering Direction\n${cumulative.steerDirection}`,
        );
      }
    }

    const history = chain.getHistory();
    if (history.length >= 2) {
      const recent = history[history.length - 1]!;
      const improvement =
        (recent.qualityAfter as number) - (recent.qualityBefore as number);
      sections.push(
        [
          '### Last Iteration Summary',
          `- Quality change: ${improvement >= 0 ? '+' : ''}${improvement.toFixed(3)}`,
          `- New patterns found: ${recent.newPatterns.length}`,
          `- Selectors failing: ${recent.failingSelectors.length}`,
        ].join('\n'),
      );
    }
  } else {
    sections.push(
      'This is the initial crawl iteration. Focus on broad extraction and discovery.',
    );
  }

  if (target) {
    const targetLines = [
      '### Current Target',
      `- URL: ${target.url}`,
      `- Spider: ${target.spiderName}`,
      `- Max pages: ${target.maxPages}`,
      `- Priority: ${target.priority}`,
    ];
    if (target.pageTypeHint) {
      targetLines.push(`- Expected page type: ${target.pageTypeHint}`);
    }
    sections.push(targetLines.join('\n'));
  }

  return sections.join('\n\n');
}
